pennylane/circuit_drawer.py

```diff
@@ -0,0 +1,916 @@
+            wire (int): The element's wire which is currently resolved
+
+        Returns:
+            str: String representation of the element
+        """
+        if element is None:
+            return ""
+        if isinstance(element, str):
+            return element
+        if isinstance(element, qml.operation.Observable) and element.return_type is not None:
+            return self.output_representation(element, wire)
+
+        return self.operator_representation(element, wire)
+
+
+def _remove_duplicates(input_list):
+    """Remove duplicate entries from a list.
+
+    This operation preserves the order of the list's elements.
+
+    Args:
+        input_list (list[Hashable]): The list whose duplicate entries shall be removed
+
+    Returns:
+        list[Hashable]: The input list without duplicate entries
+    """
+    return list(OrderedDict.fromkeys(input_list))
+
+
+class CircuitDrawer:
+    """Creates a circuit diagram from the operators of a CircuitGraph in grid form.
+
+    Args:
+        raw_operation_grid (list[list[qml.operation.Operation]]): The CircuitGraph's operations
+        raw_observable_grid (list[list[qml.operation.Observable]]): The CircuitGraph's observables
+        charset (pennylane.circuit_drawer.CharSet, optional): The CharSet that shall be used for drawing. Defaults to UnicodeCharSet.
+        show_variable_names (bool, optional): Show variable names instead of variable values. Defaults to False.
+    """
+
+    def resolve_representation(self, grid, representation_grid):
+        """Resolve the string representation of the given Grid.
+
+        Args:
+            grid (pennylane.circuit_drawer.Grid): Grid that holds the circuit information
+            representation_grid (pennylane.circuit_drawer.Grid): Grid that is used to store the string representations
+        """
+        for i in range(grid.num_layers):
+            representation_layer = [""] * grid.num_wires
+
+            for wire, operator in enumerate(grid.layer(i)):
+                representation_layer[wire] = self.representation_resolver.element_representation(
+                    operator, wire
+                )
+
+            representation_grid.append_layer(representation_layer)
+
+    def resolve_decorations(self, grid, representation_grid, separate):
+        """Resolve the decorations of the given Grid.
+
+        Args:
+            grid (pennylane.circuit_drawer.Grid): Grid that holds the circuit information
+            representation_grid (pennylane.circuit_drawer.Grid): Grid that holds the string representations and into
+                which the decorations will be inserted
+            separate (bool): Insert decorations into separate layers
+
+        Returns:
+            list[int]: List with indices of inserted decoration layers
+        """
+        j = 0
+        inserted_indices = []
+
+        for i in range(grid.num_layers):
+            layer_operators = _remove_duplicates(grid.layer(i))
```

Comment:
At this point, I'm slightly lost, how come duplicates need to be removed here? What would be duplicated for the specific ``layer``?

Reply:
How about breaking this one up somehow like as follows:
```python
def build_decoration_layer(self, decoration_layer, wires):
    sorted_wires = wires.copy()
    sorted_wires.sort()

    decoration_layer[sorted_wires[0]] = self.charset.TOP_MULTI_LINE_GATE_CONNECTOR

    for k in range(sorted_wires[0] + 1, sorted_wires[-1]):
        if k in sorted_wires:
            decoration_layer[k] = self.charset.MIDDLE_MULTI_LINE_GATE_CONNECTOR
        else:
            decoration_layer[k] = self.charset.EMPTY_MULTI_LINE_GATE_CONNECTOR

    decoration_layer[
        sorted_wires[-1]
    ] = self.charset.BOTTOM_MULTI_LINE_GATE_CONNECTOR
    return decoration_layer
```
And creating separate methods for ``separate=True`` and ``separate=False`` might also be a way to increase readability (though it might also just complicate things actually).

Reply:
Thanks for that input! I took it up and also managed to get rid of `separate` entirely.